File: c#/_87_PrimePowerTriples.cs
using System;
using System.Collections.Generic;

public class _87_PrimePowerTriples{
	public static int Euler87(){
		int max = 50000000;
        int maxSqrt = (int) Math.Sqrt(max);
        int maxCbrt = (int) Math.Cbrt(max);
        int maxFtrt = (int) Math.Sqrt(Math.Sqrt(max));
		bool[] sieved = new bool[maxSqrt];
		List<int> squares = new List<int>();
		List<int> cubes = new List<int>();
		List<int> fourths = new List<int>();
		for (int i = 2; i < sieved.Length; i++){
			if (!sieved[i]){
				squares.Add(i * i);
				if (i < maxCbrt) cubes.Add(i * i * i);
				if (i < maxFtrt) fourths.Add(i * i * i * i);
				for (int j = 2 * i; j < sieved.Length; j += i){
					sieved[j] = true;
				}
			}
		}
		HashSet<int> sums = new HashSet<int>();
		foreach (int fourth in fourths){
			foreach (int cube in cubes){
				if (fourth + cube >= max) break;
				foreach (int square in squares){
					int sum = square + cube + fourth;
					if (sum >= max) break;
					sums.Add(sum);
				}
			}
		}
		return sums.Count;
	}
	
	public static void Main(){
		Console.WriteLine(Euler87());
	}
}

File: java/_87_PrimePowerTriples.java
import java.util.*;

public class _87_PrimePowerTriples{
    
    public static void main(String[] args){
        
        int n = 50000000;
        
        ArrayList<Integer> squares = new ArrayList<>();
        ArrayList<Integer> cubes = new ArrayList<>();
        ArrayList<Integer> fourths = new ArrayList<>();
        
        int maxSqrt = (int) Math.floor(Math.sqrt(n));
        int maxCbrt = (int) Math.floor(Math.cbrt(n));
        int max4th  = (int) Math.floor(Math.sqrt(Math.sqrt(n)));
        
        boolean[] sieved = new boolean[maxSqrt];
        for (int i = 2; i < sieved.length; i++){
            if (!sieved[i]){
                for (int j = i+i; j < sieved.length; j += i){
                    sieved[j] = true;
                }
                if (i <= max4th) fourths.add(i*i*i*i);
                if (i <= maxCbrt) cubes.add(i*i*i);
                squares.add(i*i);
            }
        }
        
        HashSet<Integer> triples = new HashSet<>();
        for (int square : squares){
            for (int cube : cubes){
                for (int fourth : fourths){
                    int sum = square + cube + fourth;
                    if (sum >= n){
                        break;
                    }
                    else{
                        triples.add(sum);
                    }
                }
            }
        }
        
        System.out.println("Numbers below fifty million that can be expressed as a sum of a prime square, cube, and fourth power: " + triples.size()); // 1097343
    }
}